psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1424 @@
+                    R_iajb[ij]->add(temp);
+                }
+            }
+
+            R_iajb_rms[ij] = R_iajb[ij]->rms();
+        }
+
+        // evaluate convergence using current amplitudes and residuals
+        e_prev = e_curr;
+        e_curr = compute_iteration_energy(R_iajb);
+        r_curr = *max_element(R_iajb_rms.begin(), R_iajb_rms.end());
+
+        r_converged = (fabs(r_curr) < options_.get_double("R_CONVERGENCE"));
+        e_converged = (fabs(e_curr - e_prev) < options_.get_double("E_CONVERGENCE"));
+
+// use residuals to get next amplitudes
+#pragma omp parallel for schedule(static, 1)
+        for (int ij = 0; ij < n_lmo_pairs; ++ij) {
+            int i, j;
+            std::tie(i, j) = ij_to_i_j_[ij];
+            for (int a = 0; a < n_pno_[ij]; ++a) {
+                for (int b = 0; b < n_pno_[ij]; ++b) {
+                    T_iajb_[ij]->add(a, b, -R_iajb[ij]->get(a, b) / ((e_pno_[ij]->get(a) + e_pno_[ij]->get(b)) -
+                                                                    (F_lmo_->get(i, i) + F_lmo_->get(j, j))));
+                }
+            }
+        }
+
+        // DIIS extrapolation
+        auto T_iajb_flat = flatten_mats(T_iajb_);
+        auto R_iajb_flat = flatten_mats(R_iajb);
+        if (iteration == 0) {
+            diis.set_error_vector_size(1, DIISEntry::Matrix, R_iajb_flat.get());
+            diis.set_vector_size(1, DIISEntry::Matrix, T_iajb_flat.get());
+        }
+        diis.add_entry(R_iajb_flat, T_iajb_flat);
+        diis.extrapolate(T_iajb_flat);
+        copy_flat_mats(T_iajb_flat, T_iajb_);
+
+#pragma omp parallel for schedule(static, 1)
+        for (int ij = 0; ij < n_lmo_pairs; ++ij) {
+            Tt_iajb_[ij]->copy(T_iajb_[ij]);
+            Tt_iajb_[ij]->scale(2.0);
+            Tt_iajb_[ij]->subtract(T_iajb_[ij]->transpose());
+        }
+
+        outfile->Printf("  @LMP2 iter %3d: %16.12f %10.3e %10.3e\n", iteration, e_curr, e_curr - e_prev, r_curr);
+
+        iteration++;
+
+        if(iteration > max_iteration) {
+            throw PSIEXCEPTION("Maximum DLPNO iterations exceeded.");
+        }
+    }
+
+    e_lmp2_ = e_curr;
+}
+
+double DLPNOMP2::compute_iteration_energy(const std::vector<SharedMatrix> &R_iajb) {
+    double e_mp2 = 0.0;
+    for (int ij = 0; ij < Tt_iajb_.size(); ++ij) {
+        e_mp2 += K_iajb_[ij]->vector_dot(Tt_iajb_[ij]);
+        e_mp2 += R_iajb[ij]->vector_dot(Tt_iajb_[ij]);
+    }
+    return e_mp2;
+}
+
+void DLPNOMP2::setup() {
+    int natom = molecule_->natom();
+    int nbf = basisset_->nbf();
+    int nshell = basisset_->nshell();
+    int naux = ribasis_->nbf();
+    int nshellri = ribasis_->nshell();
+    int naocc = nalpha_ - nfrzc();
+
+    auto C_occ = reference_wavefunction_->Ca_subset("AO", "OCC");
+
+    // Localize active occupied orbitals
+    //    auto localizer = BoysLocalizer(basisset_, reference_wavefunction_->Ca_subset("AO", "ACTIVE_OCC"));
+    if (options_.get_str("DLPNO_LOCAL_ORBITALS") == "BOYS") {
```

Comment:
Is there a reason why you can't declare a localizer outside of the if statements, define the localizer in the if statements, and call the methods outside the if statements using polymorphism? I know that C++ is picky about types, but I would expect that would still run.

Reply:
You might be more familiar with polymorphism in C++. I tried something like this:
```
    Localizer localizer;
    if (options_.get_str("DLPNO_LOCAL_ORBITALS") == "BOYS") {
        localizer = BoysLocalizer(basisset_, reference_wavefunction_->Ca_subset("AO", "ACTIVE_OCC"));
    } else if (options_.get_str("DLPNO_LOCAL_ORBITALS") == "PIPEK_MEZEY") {
        localizer = PMLocalizer(basisset_, reference_wavefunction_->Ca_subset("AO", "ACTIVE_OCC"));
    }
...
```
and I get the following error:
```
error: object of abstract class type "psi::Localizer" is not allowed:
            function "psi::Localizer::print_header" is a pure virtual function
            function "psi::Localizer::localize" is a pure virtual function
```

Unless you know of some pointer magic to get around declaring an object with an abstract class, it looks like the suggested change might require some modification of the `Localizer` class.

Reply:
That's a good argument for modifying the Localizer class, but that is beyond the scope of this PR.